### PR TITLE
Add more SONIC module types

### PR DESCRIPTION
**Pull request by kpedro88**
#### PR description:

* Add `SonicEDFilter` (also a `stream` module)
* Add `SonicOneEDAnalyzer` (a `one` module, synchronous calls required since ExternalWork is not compatible)

It turns out that ExternalWork is not available for `stream` or `global` EDAnalyzer modules. This could be added in the framework, but I haven't had time to do that so far. It will be considered for a later development item, if there are motivating use cases. (The `one` analyzer module is intended to facilitate rapid prototyping by end users who want ML inference results for their analysis. For more complicated cases, it's still probably best to use `SonicEDProducer` for inference and a separate `one` analyzer just to write the ROOT files.)

Some common operations are extracted into utility functions and a common base class for modules that use ExternalWork. (The common base class re-introduces some template dependent scope issues, but these are contained within SonicCore, and should not affect downstream implementation.) Documentation and tests are updated accordingly.

#### PR validation:

Unit tests pass.

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
please test

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc [hunk 1]
```diff
@@ -29,7 +29,12 @@ void SonicClientBase::setDebugName(const std::string& debugName) {
 }
 
 void SonicClientBase::start(edm::WaitingTaskWithArenaHolder holder) {
+  start();
   holder_ = std::move(holder);
+}
+
+void SonicClientBase::start() {
+  holder_.reset();
```
Comment (makortel) [inline HeterogeneousCore/SonicCore/src/SonicClientBase.cc:37]: Is this `reset()` really needed, or is it more like a precaution?

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc [hunk 2]
```diff
@@ -57,7 +62,10 @@ void SonicClientBase::finish(bool success, std::exception_ptr eptr) {
     edm::LogInfo(fullDebugName_) << "Client time: "
                                  << std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0_).count();
   }
-  holder_.doneWaiting(eptr);
+  if (holder_)
+    holder_->doneWaiting(eptr);
+  else if (eptr)
```
Comment (makortel) [inline HeterogeneousCore/SonicCore/src/SonicClientBase.cc:67]: How about
```suggestion
  if (holder_) {
    holder_->doneWaiting(eptr);
    holder_.reset();
  } else if (eptr)
```
to make it clear that the old `WaitingTaskWithArenaHolder` won't be used e.g. by any (future) mistake?

**Review by makortel — COMMENTED** [review]
Overall looks good to me (my comments are close to nitpicking).

**Comment by kpedro88** [comment]
please test

**Comment by makortel** [comment]
I do not understand how this PR could cause tiny changes in 
* 158.01 `ParticleFlow/JetResponse/slimmedJetsPuppi/JEC/preso_eta30`
* 23234.0 `SiOuterTrackerV/Tracks/FinalResolution/d0Resolution`
* 28234.0 `SiOuterTrackerV/Tracks/FinalResolution/VtxZResolution`

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**